### PR TITLE
Support Markdown output format for verification reports

### PR DESCRIPTION
**Pull request by novotnyr**
Support Markdown output format for verification reports.

Enable `-vrf` or `-verification-reports-formats` CLI switch that enables multiple output formats: plain, HTML and Markdown. By default, Markdown is disabled.

Reports are emitted into the same directory as HTML outputs, with `report.markdown` being the default filename.

## Out of scope

Plugin Problem messages are not in Markdown. This is out of scope of this PR.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -246,6 +246,20 @@ Here is the full syntax of the command:
...
+    Supported formats are: `plain` (console output), `html` and `markdown`
+    By default, `plain` and `html` output formats are enabled.
+    Multiple output formats are supported, separated by a comma.
+
+    Output format that starts with a `-` (dash) will be suppressed: either from the default 
+    set of output formats or from the specified output formats.
+
+    For example, `plain,markdown` will enable console output and the Markdown verification reports.
+
+    Setting the `-plain` will disable console output but retain the default HTML output.
```
Comment (YannCebron) [inline README.md:261]: group the two samples as bullet polnt list:

- `value`: explanation
- `value2`: explanation 2

File: intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/options/CmdOpts.kt [hunk 2]
```diff
@@ -4,12 +4,16 @@
...
 package com.jetbrains.pluginverifier.options
 
+import com.jetbrains.pluginverifier.output.OutputFormat
 import com.sampullara.cli.Argument
 
 open class CmdOpts(
   @set:Argument("verification-reports-dir", alias = "vrd", description = "The directory where the verification report files will reside")
   var verificationReportsDir: String? = null,
 
+  @set:Argument("verification-reports-formats", alias="vrf", description = "Output format of the verification report files. Supported formats are 'stream' (stdout), 'html' and 'markdown'")
```
Comment (YannCebron) [inline intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/options/CmdOpts.kt:14]: `plain`, not `stream` for console?!

File: intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/options/OptionsParser.kt [hunk 3]
```diff
@@ -54,14 +56,44 @@ object OptionsParser {
...
 
+  private fun parseOutputFormats(opts: CmdOpts): List<OutputFormat> {
+    val (exclusions, inclusions) = opts.outputFormats.partition { it.trim().startsWith("-") }
+
+
+    val includedFormats = inclusions.mapNotNull {
+      try {
+        OutputFormat.valueOf(it.uppercase())
+      } catch (e: IllegalArgumentException) {
+        LOG.warn("Unsupported verification report output format '$it'. Skipping")
```
Comment (YannCebron) [inline intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/options/OptionsParser.kt:77]: IMHO we should fail execution instead if we can't parse args
not sure log is always visible to user

File: intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/tasks/checkPlugin/CheckPluginResultPrinter.kt [hunk 4]
```diff
@@ -24,11 +25,20 @@ class CheckPluginResultPrinter(private val pluginRepository: PluginRepository) :
...
       } else {
-        printOnStdout(this)
+        if (outputOptions.usePlainOutput()) {
+          printOnStdout (this)
+        }
       }
 
       results.groupBy { it.verificationTarget }.forEach { (verificationTarget, resultsOfIde) ->
-        HtmlResultPrinter(verificationTarget, outputOptions).printResults(resultsOfIde)
+        if(outputOptions.useHtml()) {
```
Comment (YannCebron) [inline intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/tasks/checkPlugin/CheckPluginResultPrinter.kt:34]: missing space

File: intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/tasks/twoTargets/TwoTargetsResultPrinter.kt [hunk 5]
```diff
@@ -40,8 +43,14 @@ class TwoTargetsResultPrinter : TaskResultPrinter {
         println("Enable TeamCity results printing option (-team-city or -tc) to see the results in TeamCity builds format.")
       }
 
-      HtmlResultPrinter(baseTarget, outputOptions).printResults(baseResults)
-      HtmlResultPrinter(newTarget, outputOptions).printResults(newResults)
+      if(outputOptions.useHtml()) {
```
Comment (YannCebron) [inline intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/tasks/twoTargets/TwoTargetsResultPrinter.kt:46]: missing space

File: intellij-plugin-verifier/verifier-intellij/src/main/java/com/jetbrains/pluginverifier/usages/nonExtendable/NonExtendableTypeInherited.kt [hunk 6]
```diff
@@ -14,6 +14,14 @@ import com.jetbrains.pluginverifier.results.presentation.formatClassLocation
 import com.jetbrains.pluginverifier.usages.formatUsageLocation
 import java.util.*
 
+/**
```
Comment (YannCebron) [inline intellij-plugin-verifier/verifier-intellij/src/main/java/com/jetbrains/pluginverifier/usages/nonExtendable/NonExtendableTypeInherited.kt:17]: please remove unrelated changes from PR

File: intellij-plugin-verifier/verifier-test/src/test/java/com/jetbrains/pluginverifier/output/markdown/MarkdownOutputTest.kt [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+val mockMethodLocation = MethodLocation(
+  ClassLocation("SomeClass", null, Modifiers.of(PUBLIC), SomeFileOrigin),
+  "someMethod",
+  "()V",
+  emptyList(),
+  null,
+  Modifiers.of(PUBLIC)
+)
+
+private val violetClassLocation = ClassLocation("VioletClass", null, Modifiers.of(PUBLIC), SomeFileOrigin)
```
Comment (YannCebron) [inline intellij-plugin-verifier/verifier-test/src/test/java/com/jetbrains/pluginverifier/output/markdown/MarkdownOutputTest.kt:346]: what does "violet" mean here?

File: intellij-plugin-verifier/verifier-test/src/test/java/com/jetbrains/pluginverifier/output/markdown/MarkdownOutputTest.kt [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+        Plugin can probably be enabled or disabled without IDE restart
+
+
+      """.trimIndent()
+    assertEquals(expected, output())
+  }
+
+  private fun output() = out.buffer.toString()
+}
+
```
Comment (YannCebron) [inline intellij-plugin-verifier/verifier-test/src/test/java/com/jetbrains/pluginverifier/output/markdown/MarkdownOutputTest.kt:316]: AFAIU everything below should have `private` modifier

**Review by YannCebron — CHANGES_REQUESTED** [review]

File: intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/options/OptionsParser.kt (hunk 3 above)
Comment (novotnyr) [reply intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/options/OptionsParser.kt:77]: There is always a fallback on default format to maintain backwards compatibility, even in case of failures or syntax errors. This message just logs the fallback.

Furthermore, the log is always visible and enabled for `cli`. This is provided by [explicit configuration of Logback](https://github.com/JetBrains/intellij-plugin-verifier/blob/master/intellij-plugin-verifier/verifier-cli/src/main/resources/logback.xml).

File: intellij-plugin-verifier/verifier-test/src/test/java/com/jetbrains/pluginverifier/output/markdown/MarkdownOutputTest.kt (hunk 7 above)
Comment (novotnyr) [reply intellij-plugin-verifier/verifier-test/src/test/java/com/jetbrains/pluginverifier/output/markdown/MarkdownOutputTest.kt:346]: Using more generic nomenclature.

File: intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/options/CmdOpts.kt (hunk 2 above)
Comment (novotnyr) [reply intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/options/CmdOpts.kt:14]: Unified to `plain`.

**Review by YannCebron — APPROVED** [review]

**Review by YannCebron — APPROVED** [review]

File: intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/tasks/checkPlugin/CheckPluginResultPrinter.kt [hunk 9]
```diff
@@ -24,11 +27,28 @@ class CheckPluginResultPrinter(private val pluginRepository: PluginRepository) :
         val teamCityHistory = printTcLog(true, outputOptions.teamCityLog, outputOptions.teamCityGroupType)
         outputOptions.postProcessTeamCityTests(teamCityHistory)
       } else {
-        printOnStdout(this)
+        if (outputOptions.usePlainOutput()) {
+          printOnStdout (this)
+        }
+      }
+
+      if (hasOnlyPluginsWithInvalidFiles()) {
```
Comment (LChernigovskaya) [inline intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/tasks/checkPlugin/CheckPluginResultPrinter.kt:35]: Comment on why MarkdownResultPrinter without a markdown option is used here may be useful.

**Review by LChernigovskaya — APPROVED** [review]

File: intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/tasks/checkPlugin/CheckPluginResultPrinter.kt (hunk 9 above)
Comment (novotnyr) [reply intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/tasks/checkPlugin/CheckPluginResultPrinter.kt:35]: Thanks for spotting this. Markdown should be used only when explicitly requested. Fixed.

File: intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/tasks/checkPlugin/CheckPluginResultPrinter.kt [hunk 10]
```diff
@@ -34,8 +34,10 @@ class CheckPluginResultPrinter(private val pluginRepository: PluginRepository) :
 
       if (hasOnlyPluginsWithInvalidFiles()) {
         for ((ideDescriptor, invalidPlugins) in ideDescriptorsWithInvalidPlugins) {
-          MarkdownResultPrinter.create(ideDescriptor.toVerificationTarget(), outputOptions).use {
-            it.printInvalidPluginFiles(invalidPlugins)
+          if(outputOptions.useMarkdown()) {
```
Comment (YannCebron) [inline intellij-plugin-verifier/verifier-cli/src/main/java/com/jetbrains/pluginverifier/tasks/checkPlugin/CheckPluginResultPrinter.kt:37]: missing space before `(`

**Review by YannCebron — CHANGES_REQUESTED** [review]

**Review by YannCebron — APPROVED** [review]
